Add additional MDAOs for intervals and cron script to roll up PMs.

jlhughes opened the pull request:
Additional pmInfoDAOs (MDAO) added for 5 minute, 1 hour, 1 day.
Cron script running every minute replaces the pmInfoDAO, and copies
the contents into the other interval DAOs.
The 5 minute, 1 hour, 1 day DAOs are cleared every 5 minutes, 1 hour, 1 day respectively.

Added menu entry for Performance - 1 Day.

The interval MDAOs are intended to be used with the future TimeSeriesDAO.

jlhughes committed:
Add additional MDAOs for intervals and cron script to roll up PMs.

Additional pmInfoDAOs (MDAO) added for 5 minute, 1 hour, 1 day.
Cron script running every minute replaces the pmInfoDAO, and copies
the contents into the other interval DAOs.
The 5 minute, 1 hour, 1 day DAOs are cleared every 5 minutes, 1 hour, 1 day respectively.

Added menu entry for Performance - 1 Day.

The interval MDAOs are intended to be used with the future TimeSeriesDAO.

TW80000 commented on src/foam/nanos/pm/DAOPMLoggerCopySink.java:
```diff
@@ -0,0 +1,51 @@
+    }
+    PMInfo info = (PMInfo) obj;
+    for ( DAO dao : toDAOs_ ) {
+      PMInfo i = (PMInfo) dao.find(info);
```
Doesn't this need `.inX(x)` to make sure we keep using the user's context when accessing the other DAOs?

TW80000 commented on src/foam/nanos/pm/DAOPMLoggerCopySink.java:
```diff
@@ -0,0 +1,51 @@
+import foam.dao.ProxySink;
+
+/**
+ * Copy an PMInfoDAO entries into other PMInfoDAOs.
```
Typo I think. Assuming "an" should be "any".

TW80000 commented on src/foam/nanos/pm/cronjobs.jrl:
```diff
@@ -0,0 +1 @@
+p({"class":"foam.nanos.cron.Cron","id":"PMInfoRollup","enabled":true,code":"import foam.dao.DAO;\nimport foam.nanos.logger.Logger;\nimport foam.nanos.pm.PM;\nimport foam.nanos.pm.PMInfo;\nimport foam.nanos.pm.DAOPMLoggerCopySink;\nimport foam.nanos.script.Script;\n\nscriptName = \"PMInfoRollup\";\npm = new PM(Script.class, \"scriptName\");\nlogger = x.get(\"logger\");\nlogger.info(new String[] {s ...
```
There's a syntax error in here. Looks like a problem with double quote characters not being escaped properly.

```
2019.04.29.09.09.31.005,SEVERE,[JDAO],cronjobs.0,Parse error,Invalid character '"' found at 71
Valid characters include: 'N','k','J','g','F','c','B','_','[','x','W','6','t','S','2','p','O','l','K','h','G','d','C','`','\','y','X','7','u','T','3','q','P','m','L','i','H','e','D','a',']','z','Y','8','v','U','4','r','Q','0','n','M','j','I','f','E','$','b','A','^','Z','9','w','V','5','s','R','1','o',line:,{"class":"foam.nanos.cron.Cron","id":"PMInfoRollup","enabled":true,code":"import foam.dao.DAO;\nimport foam.nanos.logger.Logger;\nimport foam.nanos.pm.PM;\nimport foam.nanos.pm.PMInfo;\nimport foam.nanos.pm.DAOPMLoggerCopySink;\nimport foam.nanos.script.Script;\n\nscriptName = \"PMInfoRollup\";\npm = new PM(Script.class, \"scriptName\");\nlogger = x.get(\"logger\");\nlogger.info(new String[] {scriptName, \"start\"});\n\n// current 1 minute dao\ndao = x.get(\"pmInfoDAO\").getDelegate();\n\n// replace with new empty dao\nx.get(\"pmInfoDAO\").setDelegate(new foam.dao.MDAO(foam.nanos.pm.PMInfo.getOwnClassInfo()));\n\none = x.get(\"pmInfo1minDAO\");\nfive = x.get(\"pmInfo5minDAO\");\nhour = x.get(\"pmInfo1hourDAO\");\nday = x.get(\"pmInfo1dayDAO\");\n\n//logger.info(new String[] {scriptName, \"daos\", one.toString(), five.toString(), hour.toString(), day.toString()});\n//logger.info(new String[] {scriptName, \"current replaced\"});\n\n// move previous 1 minute to named 1 minute dao\none.setDelegate(dao);\n\n// cleanup\nnow = java.time.LocalDateTime.now();\n\n// if at 5 minute interval of hour clear 5 minute dao\nif ( now.getMinute() % 5 == 0 ) {\n  five.removeAll();\n  logger.info(new String[] {scriptName,\"clean 5 minute\"});\n}  \n\n// if at top of the hour, clear hour dao, if start of day.\nif ( now.getMinute() == 0 ) {\n  hour.removeAll();\n  logger.info(new String[] {scriptName,\"clean 1 hour\"});\n  if ( now.getHour() == 0 ) {\n    day.removeAll();\n    logger.info(new String[] {scriptName,\"clean 1 day\"});\n  }\n}\n\n// put 1 minute entries into other daos.\nDAO[] daos = new DAO[] { five, hour, day };\ndao.select(new DAOPMLoggerCopySink(x, daos));\nlogger.info(new String[] {scriptName,\"end\"});\npm.log(x);\n","output":""}
```

TW80000 commented on src/foam/nanos/pm/services.jrl:
```diff
@@ -0,0 +1,7 @@
+p({"class":"foam.nanos.boot.NSpec", "name":"pmLogger",                         "lazy":true,  "serve":false, "serviceClass":"foam.nanos.pm.DAOPMLogger"})
+p({"class":"foam.nanos.boot.NSpec", "name":"pmInfoDAO",                        "lazy":true,  "serve":true,  "serviceScript":"return new foam.dao.ProxyDAO.Builder(x).setDelegate(new foam.dao.MDAO(foam.nanos.pm.PMInfo.getOwnClassInfo())).build();","client":"{\"of\":\"foam.nanos.pm.PMInfo\", \"remoteListenerSupport\": true}" })
```
Why do all of these services have a plain `ProxyDAO` decorating them?

jlhughes replied:
I really only need the plain ProxyDAO on the pmInfoDAO. I use it to replace the delegate during rollup.  After that it was cut and paste.

TW80000 commented on src/foam/nanos/pm/services.jrl:
```diff
@@ -0,0 +1,7 @@
+p({"class":"foam.nanos.boot.NSpec", "name":"pmLogger",                         "lazy":true,  "serve":false, "serviceClass":"foam.nanos.pm.DAOPMLogger"})
+p({"class":"foam.nanos.boot.NSpec", "name":"pmInfoDAO",                        "lazy":true,  "serve":true,  "serviceScript":"return new foam.dao.ProxyDAO.Builder(x).setDelegate(new foam.dao.MDAO(foam.nanos.pm.PMInfo.getOwnClassInfo())).build();","client":"{\"of\":\"foam.nanos.pm.PMInfo\", \"remoteListenerSupport\": true}" })
```
These services should probably be behind some authorization checks so they're not open to everyone. Wrapping in `AuthenticatedDAO` should be sufficient.

TW80000 reviewed (commented):
Wasn't able to review the cron job since it didn't get parsed correctly, but will review that when it's fixed.

jlhughes committed:
Address code review comments

- use inX on dao find
- remove uncessary ProxyDAOs
- add missing quote in cronjob script

jlhughes commented:
superseded by Candlestick for PMs
